Integrating block-sync in an init-container

kushalShukla-web opened the pull request:
1. Changes that are introduced in this pr , followed by  #760 .
   Integrated `download-key `in an init container that downloads the `key.yml` file from the         Prometheus repository. It uses the `key.sh` script to fetch the pull request and copy the `key.yml` file into the volumeMount section:
```
volumeMounts:
  - name: key
    mountPath: /config

```
2. Integrated `data-downloader`on both pr and released side, which uses the `block-sync` binary to download `pre-generated` block data.
3. Updated the `prometheus-builder` image, modifying the `builder.sh` file to copy the `key.yml` file into the volumeMount section so that the PR section can also download the pre-generated block data.

kushalShukla-web committed:
Added yaml file changes and prometheus builder changes that was removed
from block-sync.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bboreham commented on prombench/docs/kind.md:
````diff
@@ -114,7 +114,20 @@ If used with the GitHub integration:
+> Note: Make sure this secret applied before ```3b_prometheus-test_deployment.yaml```
+- **Option 2: Skip downloading data**
+
+   If you don’t need to download data, edit the `3b_prometheus-test_deployment.yaml` file:
````
Could we instead tell people to create an empty Secret?

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -34,7 +34,7 @@ spec:
       initContainers:
       - name: prometheus-builder
-        image: docker.io/prominfra/prometheus-builder:master
+        image: kushalshukla/builder
```
Fix up before merge.

bboreham commented on tools/prometheus-builder/build.sh:
```diff
@@ -6,7 +6,7 @@ if [[ -z $PR_NUMBER || -z $VOLUME_DIR || -z $GITHUB_ORG || -z $GITHUB_REPO ]]; t
     exit 1;
 fi
-
+ 
```
Spurious change

bboreham commented on tools/prometheus-builder/Dockerfile:
```diff
@@ -4,6 +4,8 @@ RUN mkdir -p /go/src/github.com
 RUN chmod +x /go/src/github.com/build.sh
 
-ENTRYPOINT ["/go/src/github.com/build.sh"]
+ENTRYPOINT ["/go/src/github.com/build.sh"]
```
Spurious change

bboreham commented on tools/prometheus-builder/Dockerfile:
```diff
@@ -4,6 +4,8 @@ RUN mkdir -p /go/src/github.com
 
 COPY ./build.sh /go/src/github.com/build.sh
 
+COPY ./key.sh /download-key/key.sh
```
Why is this in such a different place from the other script?

bboreham commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,11 +24,22 @@ fi
+# emptyDir. This file will later be used by the data-downloader init container.
+MKDIR="/config"
+if [ -f "$DIR/key.yml" ]; then
+    echo "File exists."
```
Who is expected to read this message, and what are they expected to do about it?

bboreham commented on prombench/docs/kind.md:
```diff
@@ -114,7 +114,20 @@ If used with the GitHub integration:
 
-2. Deploy the Kubernetes objects:
+2. Before applying benchmarking objects , You have two choices to make:  
+- **Option 1: Download data from object storage**
```
How would someone coming to this for the first time understand this choice?
Can we give some context why data might be downloaded?

bboreham reviewed (commented):
I'm struggling to find my way around.  Some trivial points stand out, but I need better orientation.

>  removed from block-sync:

Please explain each PR in terms of what it does, not what some other PR didn't do.

For example, you could describe the intended purpose of `key.sh`.

kakkoyun commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,6 +24,17 @@ fi
+# Here, MKDIR is specified in the volumeMount section of the prometheus-builder init container, 
+# where it will copy the key.yml file from the Prometheus directory to the volume section of the
+# emptyDir. This file will later be used by the data-downloader init container.
+MKDIR="/config"
```
I think you need to change the name.

kakkoyun commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,6 +24,17 @@ fi
 
 git checkout pr-branch
 
+# Here, MKDIR is specified in the volumeMount section of the prometheus-builder init container, 
```
It would be better to have this in another script.

kushalShukla-web committed:
1. Changed Image name
2. Modified Prometheus deployment yml file
3. changed Prometheus Builder Docker file
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Added empty secretes ,modified newstore function to normal stop if config file is empty.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Added --storage.tsdb.retention.size to retain data for certain size.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kakkoyun commented on prombench/docs/eks.md:
````diff
@@ -107,8 +107,37 @@ Run Prombench tests in [Elastic Kubernetes Service (EKS)](https://aws.amazon.com
     make node_create
     ```
+3. **Setting Up Benchmarking Data**    
+ When setting up a benchmarking environment, it’s often useful to have pre-generated data available.This data can help speed up testing and make benchmarks more realistic by simulating actual workloads.
````
```suggestion
 When setting up a benchmarking environment, it’s often useful to have pre-generated data available. This data can help speed up testing and make benchmarks more realistic by simulating actual workloads.
```

kakkoyun commented on prombench/docs/eks.md:
````diff
@@ -107,8 +107,37 @@ Run Prombench tests in [Elastic Kubernetes Service (EKS)](https://aws.amazon.com
+Here’s how each option works:
+- **Option 1: Download data from object storage**
+
+   To download data from object storage, create a Kubernetes secret with exact named `bucket-secret` and file name ```object-config.yml```  with the necessary credentials as per your object storage. This secret enables access to the stored data.
````
```suggestion
   To download data from object storage, create a Kubernetes secret with exact named `bucket-secret` and file name `object-config.yml`  with the necessary credentials as per your object storage. This secret enables access to the stored data.
```

kakkoyun commented on prombench/docs/eks.md:
````diff
@@ -107,8 +107,37 @@ Run Prombench tests in [Elastic Kubernetes Service (EKS)](https://aws.amazon.com
+- **Option 1: Download data from object storage**
+
+   To download data from object storage, create a Kubernetes secret with exact named `bucket-secret` and file name ```object-config.yml```  with the necessary credentials as per your object storage. This secret enables access to the stored data.
+> Note: Make sure this secret applied before ```3b_prometheus-test_deployment.yaml```
````
```suggestion
> Note: Make sure this secret applied before `3b_prometheus-test_deployment.yaml`
```

kakkoyun commented on prombench/docs/eks.md:
```diff
@@ -107,8 +107,37 @@ Run Prombench tests in [Elastic Kubernetes Service (EKS)](https://aws.amazon.com
+type: Opaque
+stringData:
+  object-config.yml: 
+ 
```
```suggestion
It is missing "```"
 
```

kakkoyun commented on prombench/docs/gke.md:
```diff
@@ -96,8 +96,38 @@ Run Prombench tests in [Google Kubernetes Engine (GKE)](https://cloud.google.com
+  object-config.yml: 
+ 
+Regardless of the option chosen, data stored in Prometheus will only be retained based on the configured retention settings (`--storage.tsdb.retention.size`). 
+
```
```suggestion
ditto '```'
```

kakkoyun commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,6 +24,16 @@ fi
+    echo "INFO:: key.yml file is Present on $DIR/key.yml directory."
+    cp  "$DIR/key.yml" "$STORAGE/key.yml"
+else
+    echo "INFO:: key.yml File does not exist on $DIR/key.yml directory."
```
We should add more context or remove this line. e.g. File does not exists so we won't download XXX

kakkoyun commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,6 +24,16 @@ fi
+# where it will copy the key.yml file from the Prometheus directory to the volume section of the
+# emptyDir. This file will later be used by the data-downloader init container.
+if [ -f "$DIR/key.yml" ]; then
+    echo "INFO:: key.yml file is Present on $DIR/key.yml directory."
```
"It exists so we will download the blocks from XXX."

kushalShukla-web committed:
minior changes
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Removed Image Building from config.yml file.
Removed the code for exiting container when secret is not present.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Replaced key.yml with object-config.yml file.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kakkoyun commented on tools/prometheus-builder/key.sh:
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
```
Let's be more specific concerning the filename. `key.sh` is too generic.

kakkoyun commented on tools/prometheus-builder/Dockerfile:
```diff
@@ -4,6 +4,8 @@ RUN mkdir -p /go/src/github.com
-RUN chmod +x /go/src/github.com/build.sh
+COPY ./key.sh /go/src/github.com/key.sh
+
+RUN chmod +x /go/src/github.com/*
```
We should just give executable permission in this directory. This could create security issues. Just give all the known script names.

kakkoyun reviewed (changes requested):
Looking good. I have some comments.

Also if we merge https://github.com/prometheus/test-infra/pull/810, we will have conflicts make sure we have the relevant changes specific to the PRs.

kakkoyun commented:
This needs a rebase @kushalShukla-web

kushalShukla-web commented:
Hi @kakkoyun @bboreham,  we have moved the key into the ConfigMap instead of pulling it from Prometheus, so maybe we can close this pull request ? thats why i dont rebase this.
